ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java

```diff
@@ -636,14 +639,44 @@ private void handleAdminRequest(Request request) throws InterruptedException, IO
               error = ServerErrorCode.Bad_Request;
             }
           }
+          response = new AdminResponse(adminRequest.getCorrelationId(), adminRequest.getClientId(), error);
+          break;
+        case CatchupStatus:
+          error = ServerErrorCode.No_Error;
+          metrics.catchupStatusRequestQueueTimeInMs.update(requestQueueTime);
+          metrics.catchupStatusRequestRate.mark();
+          processingTimeHistogram = metrics.catchupStatusResponseQueueTimeInMs;
+          responseQueueTimeHistogram = metrics.catchupStatusResponseQueueTimeInMs;
+          responseSendTimeHistogram = metrics.catchupStatusResponseSendTimeInMs;
+          requestTotalTimeHistogram = metrics.catchupStatusRequestTotalTimeInMs;
+          CatchupStatusAdminRequest catchupStatusRequest =
+              CatchupStatusAdminRequest.readFrom(requestStream, adminRequest);
+          if (catchupStatusRequest.getPartitionId() != null) {
+            error = validateRequest(catchupStatusRequest.getPartitionId(), RequestOrResponseType.AdminRequest);
+            partitionIds = Collections.singletonList(catchupStatusRequest.getPartitionId());
+          } else {
+            partitionIds = clusterMap.getAllPartitionIds();
+          }
+          boolean isCaughtUp = false;
+          if (!error.equals(ServerErrorCode.Partition_Unknown)) {
+            error = ServerErrorCode.No_Error;
+            isCaughtUp = isAllRemoteLagLesserOrEqual(partitionIds, catchupStatusRequest.getAcceptableLagInBytes());
+          }
+          AdminResponse adminResponse =
+              new AdminResponse(adminRequest.getCorrelationId(), adminRequest.getClientId(), error);
+          response = new CatchupStatusAdminResponse(isCaughtUp, adminResponse);
           break;
       }
-      response = new AdminResponse(adminRequest.getCorrelationId(), adminRequest.getClientId(), error);
     } catch (Exception e) {
       logger.error("Unknown exception for admin request {}", adminRequest, e);
+      metrics.unExpectedAdminOperationError.inc();
       response =
           new AdminResponse(adminRequest.getCorrelationId(), adminRequest.getClientId(), ServerErrorCode.Unknown_Error);
-      metrics.unExpectedAdminOperationError.inc();
+      switch (adminRequest.getType()) {
+        case CatchupStatus:
+          response = new CatchupStatusAdminResponse(false, response);
```

Comment:
This is to make it easier to deserialize the response? i.e. you won't have to check the error code to decide whether to deserialize the rest.

Reply:
This is because the invoker of `CatchupStatusAdminRequest` is expected a `CatchupStatusAdminResponse` and not a vanilla `AdminResponse`.